network/common/src/main/java/org/apache/spark/network/buffer/WrappedLargeByteBuffer.java

```diff
@@ -0,0 +1,252 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package org.apache.spark.network.buffer;
+
+import sun.nio.ch.DirectBuffer;
+
+import java.io.IOException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.Arrays;
+import java.util.List;
+
+public class WrappedLargeByteBuffer implements LargeByteBuffer {
+
+  //only public for tests
+  public final ByteBuffer[] underlying;
+
+  private final long size;
+  private long _pos;
+  private int currentBufferIdx;
+  private ByteBuffer currentBuffer;
+
+
+  public WrappedLargeByteBuffer(ByteBuffer[] underlying) {
+    this(underlying, findExpectedInitialPosition(underlying));
+  }
+
+  private static long findExpectedInitialPosition(ByteBuffer[] bufs) {
```

Comment:
I'm a little confused about what this is supposed to be and what the checks are meant to check... could you write a comment explaining all that?

I have a feeling that this might be over-complicating something that is supposed to be simple.